yo/app/scripts/directives/copy-or-migrate-mappings.js

```diff
@@ -0,0 +1,176 @@
+
+        $scope.alerts = [];
+
+        $scope.closeAlert = function(idx) {
+            if (angular.isDefined(idx) && $scope.alerts.length > idx) {
+                $scope.alerts.splice(idx, 1);
+            }
+        };
+
+        $scope.dataModels = dataModels;
+        $scope.projects = projects;
+
+        $scope.selectedDataModel = [];
+        $scope.selectedProject = [];
+
+        var isNewProjectsHasBeenCreated = false;
+
+        $scope.updateGridData = function() {
+
+            DataModelResource.query({ format : 'medium' }, function(results) {
+
+                $scope.dataModels = loDash.filter(results, 'data_resource');
+
+            }, function() {
+                $scope.dataModels = '';
+            });
+
+            ProjectResource.query({ format : 'short' }, function(projects) {
+
+                $scope.projects = projects;
+
+            }, function() {
+                $scope.projects = '';
+            });
+
+        };
+
+        $scope.dataModelGridContent = {
+            data: 'dataModels',
+            columnDefs: [
+                {field: 'uuid', displayName: 'Id'},
+                {field: 'name', displayName: 'Name'},
+                {field: 'description', displayName: 'Description '},
+                {field: 'schema.name', displayName: 'Schema name'}
+            ],
+            enableColumnResize: false,
+            selectedItems: $scope.selectedDataModel,
+            multiSelect: false
+        };
+
+        $scope.projectGridContent = {
+            data: 'projects',
+            columnDefs: [
+                {field: 'uuid', displayName: 'Id'},
+                {field: 'name', displayName: 'Name'},
+                {field: 'description', displayName: 'Description '}
+            ],
+            enableColumnResize: false,
+            selectedItems: $scope.selectedProject,
+            multiSelect: false
+        };
+
+        function createMigrationStatusMessage(inputDataModelId, referenceProjectId, type) {
+            var actionMsg = type;
+            var itemMsg = 'mappings from reference project ' + referenceProjectId + ' to a new project';
+            var additionalItemMsg = '';
+
+            if(inputDataModelId) {
+
+                additionalItemMsg = 'with help of the schema from input data model ' + inputDataModelId;
+            }
+
+            return function(statusMsg) {
+                return [actionMsg, itemMsg, additionalItemMsg, statusMsg].join(' ');
+            };
+        }
+
+        $scope.copyMappings = function() {
+
+            var selectedProject = $scope.selectedProject[0];
+            var referenceProjectId = selectedProject.uuid;
+
+            var payload = {
+                reference_project: referenceProjectId
+            };
+
+            var finishMessage = createMigrationStatusMessage(null, referenceProjectId, 'Copying');
+
+            ngProgress.start();
+
+            var dataModelIdPromise = $q.defer();
```

Comment:
Now looking at the code, we could use the resource promise instead of having to define our own.

```javascript
ProjectResource.get({id: referenceProjectId}.$promise.then(function(project) {
  return project.input_data_model.uuid;
}).then(function(dataModelId) {
  // …
}, // …
);
```